debian/postinst

```diff
@@ -28,6 +28,12 @@ insert_nss_entry() {
     ' /etc/nsswitch.conf
 }
 
+enable_pam_mkhomedir() {
+    log "Enabling PAM mkhomedir module..."
+
+    pam-auth-update --package --enable mkhomedir
+}
+
```

Comment:
Well, the problem with that approach is that we'd need a reference counting that should be managed by the distro, otherwise we may have not set that, but some other may have called this.

So, IHMO, the best way is actually don't touch it using `pam-auth-update` but instead, include it in our `authd.in` (in fact it was how I did personally), so that it's always there when authd is installed, but we purge it on removal.

If someone else enabled it via `pam-auth-update`, it's not a problem since it will stay, and having it twice in a pam stack it's not either a problem because calling it twice it's a no-op.

So, I'd say... Just add it to our config.

Reply:
We are definitely innovating here 😅

Reply:
I wouldn't rely on `authd.in` to enable the mkhomedir module. It's either something that the user is not aware it needs enabling (most of the cases) or, if the user is aware of it, the user will definitely know how to control it. Calling `pam-auth-update` on a module that's already enabled won't add it twice, so it's safer than adding it in our config file and end up "messing" the pam configuration.

Reply:
Agreed. I still don't quite understand why this module isn't enabled by default (since it does check whether the user has a home already or not, so no issues if the homedir was created by other means), but there should not be any problems to leave it enabled after the package is purged.

Reply:
> Calling pam-auth-update on a module that's already enabled won't add it twice, so it's safer than adding it in our config file and end up "messing" the pam configuration.

It's not messing, it's just ensuring we've a dependency happening. As said, it's not a problem duplicating the module in the list.

For sure we need to add it early in authd-gdm, so it will duplicate anyways.

And I think that if we believe that mkhomedir is a dependency of us, then it *has* to be in the pam config when authd is enabled, and anybody should be able to disable it via `pam-auth-update` unless also `authd` is disabled.

Reply:
It's not a dependency. Everything, at least on our side, works without it. A lot of other stuff rely on the user having the homedir and, from looking back at `aad-auth`, a lot of users don't know that they have to enable this in order to automatically create the home directory.

Reply:
> For sure we need to add it early in authd-gdm, so it will duplicate anyways.
 
 How so? Isn't GDM a separate stack?

Reply:
> How so? Isn't GDM a separate stack?

It is, but then *also* loads the default one, it's just an addition to the base stack:

```
session [success=ok ignore=ignore module_unknown=ignore default=bad]        pam_selinux.so open
session optional        pam_keyinit.so force revoke
session required        pam_limits.so
session required        pam_env.so readenv=1
session required        pam_env.so readenv=1 envfile=/etc/default/locale
@include common-session
session optional        pam_gnome_keyring.so auto_start
```

It's quite common for GDM stack to not care about what's next, for example for smartcard we do load multiple times `pam_sss.so` and that's not an issue at all, since all the modules should support re-entrance.

> It's not a dependency. Everything, at least on our side, works without it.

Well it's kinda is for gdm, as said, since without an HOME at least ibus doesn't work, so basically a11y it's broken. And I can imagine many other components we launch on startup just refuse to work silently. I mean it's like having a guest-session, which is ok, but if not a dependency I would call it a requirement.

Reply:
Also the main issue for me with using `pam-auth-update` at package level is that we're messing with system configuration at our level, leaving the system in a state where someone has configured a module like if it was an user, but it was actually a package. And not what packages should normally do.

Leaving the admins in a position that they can disable it again, leading to a "not-perfect" experience in authd, which can lead to bug reports on us.

Reply:
I feel it's half-backed, but it's an option... In the sense, if the home is not required by shell `login` users, then we can leave them to enable the module.

But maybe there will be still complains?

My worry on `pam-auth-config` usage is that we're stepping on the feet of other packages, in fact as per the [man](https://manpages.ubuntu.com/manpages/xenial/man8/pam-auth-update.8.html):

> Packages providing PAM modules register their profiles at
       install time by calling pam-auth-update --package

And we're definitely *not* providing ourself `mkhomedir`, so it's not what we should do (and I feel that SRU team may be also not liking it).

Reply:
As I've said: we need to remember our public here... If you take a look at some of the issues we had in [aad-auth](https://github.com/ubuntu/aad-auth/issues), there were quite a few issues related to the user not enabling `mkhomedir` (and it was listed in the documentation). I know that running `pam-auth-update` on a module that's not ours is not exactly ideal from our default POV (but if you look at SSSD, they also don't enable `mkhomedir` in their configuration files). It's a very complicated decision on how to enable this, I agree, but I think that enabling it through the config file is more prone to errors and ordering issues, so I would take the safe approach instead (even though not ideal, none of them are)

Reply:
Sure, sure it does work for that reason.

But the reason why I said to add it there too, it was to sure that's used before the gnome keyring module